src/crud/session.py

```diff
@@ -685,6 +727,41 @@ async def set_peer_config(
     result = await db.execute(stmt)
     session_peer = result.scalar_one_or_none()
 
+    # Check if this update would exceed observer limits
+    if config.observe_others:
+        # Check if peer is already an observer
+        is_currently_observer = (
+            session_peer.configuration.get("observe_others", False)
+            if session_peer and session_peer.configuration
+            else False
+        )
+
+        # Only need to check limit if peer is becoming a new observer
+        if not is_currently_observer:
+            # Use a single efficient query to count existing observers
+            existing_observers_stmt = select(func.count()).where(
+                models.SessionPeer.session_name == session_name,
+                models.SessionPeer.workspace_name == workspace_name,
+                models.SessionPeer.left_at.is_(None),  # Only active peers
+                models.SessionPeer.peer_name
+                != peer_name,  # Exclude the peer being updated
+                models.SessionPeer.configuration["observe_others"].astext.cast(
+                    Boolean
+                ),  # Only observers
+            )
+            result = await db.execute(existing_observers_stmt)
+            observer_count = result.scalar() or 0
+
+            # Add one for this peer becoming an observer
+            observer_count += 1
+
+            if observer_count > settings.SESSION_OBSERVERS_LIMIT:
+                raise ValidationException(
+                    f"Cannot update peer {peer_name} configuration to observe others. "
+                    + f"Session {session_name} would have {observer_count} observers. "
+                    + f"Maximum allowed is {settings.SESSION_OBSERVERS_LIMIT} observers per session."
+                )
```

Comment:
Would it be possible to move at least this exception to a single place in exceptions.py or elsewhere, given it's used 3 times? Maybe even deduplicate some of the logic and the sql query?